test/accelerators/NNPA/numerical/CMakeLists.txt

```diff
@@ -0,0 +1,80 @@
+# SPDX-License-Identifier: Apache-2.0
+
+set(ONNX_NUMERICALTEST_SRC_DIR ${ONNX_MLIR_SRC_ROOT}/test/numerical)
+set(ONNX_NUMERICALTEST_BIN_DIR ${ONNX_MLIR_BIN_ROOT}/test/numerical)
+
+add_custom_target(numerical-nnpa)
+set_target_properties(numerical-nnpa PROPERTIES FOLDER "Tests")
+
+set(ZDNN_LIB "${ONNX_MLIR_BIN_ROOT}/src/Accelerators/NNPA/zDNN/src/zdnn/zdnn/lib/libzdnn.so")
+
+add_custom_target(check-onnx-numerical-nnpa
+  COMMENT "Running the ONNX-MLIR numerical regression tests for NNPA"
+  COMMAND "${CMAKE_CTEST_COMMAND}" -L numerical-nnpa --output-on-failure -C $<CONFIG> --force-new-ctest-process
+  USES_TERMINAL
+  DEPENDS numerical-nnpa
+  )
+set_target_properties(check-onnx-numerical-nnpa PROPERTIES FOLDER "Tests")
+# Exclude the target from the default VS build
+set_target_properties(check-onnx-numerical-nnpa PROPERTIES EXCLUDE_FROM_DEFAULT_BUILD ON)
+
+# add_numerical_unittest(test_name sources... options...
+#   This function (generally) has the same semantic as add_onnx_mlir_executable.
+#   A test with test_name is added as a ctest to the numerical testsuite and
+#   all the rest of the arguments are passed directly to add_onnx_mlir_executable.
+#   The function usage is meant to look like a call to add_onnx_mlir_executable
+#   for readability.
+#   )
+function(add_numerical_unittest test_name)
+  add_onnx_mlir_executable(${test_name} NO_INSTALL ${ARGN})
+
+  add_dependencies(numerical-nnpa ${test_name})
+  get_target_property(test_suite_folder numerical-nnpa FOLDER)
+  if (test_suite_folder)
+    set_property(TARGET ${test_name} PROPERTY FOLDER "${test_suite_folder}")
+  endif ()
+
+  # Optimization level set by ONNX_MLIR_TEST_OPTLEVEL, defaults to 3
+  add_test(NAME ${test_name}
+    COMMAND ${test_name} -O${ONNX_MLIR_TEST_OPTLEVEL} --maccel=NNPA
+    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+    )
+  set_tests_properties(${test_name} PROPERTIES LABELS numerical-nnpa)
+  set_tests_properties(${test_name} PROPERTIES ENVIRONMENT "TEST_ATOL=0.01;TEST_RTOL=0.05};LD_PRELOAD=${ZDNN_LIB}")
```

Comment:
Do we still need `LD_PRELOAD`?

Reply:
According to @cjvolzka , he said that his testing on LoZ indicated that preload were not needed anymore and that the new version of including the init within the evaluation of the graphs worked well. So I don't think its needed anymore

Reply:
I think so, too. However, just deleting `LD_PRELOAD` didn't work because zdnn_init was not found. I'm now investigating the reason.

Reply:
It's difficult to tell what went wrong without seeing some kind of error output. For tests that are driven by ctest, it helps to pick a single test and run it with verbose on to get a better idea what's happening, for example,
```
ctest -R TestGemm-NNPA --verbose
```

Reply:
I narrowed down the situation. It seems that zdnn_init() is not found when all operations run on CPU. For example in a test for single LSTM op, when using peephole, zdnn_init() is not found.  (the LSTM op runs on CPU since zdnn does not support peephole). In this case, libzdnn.a is not linked in generated so file, `TestLSTM-NNPA_main_graph.so`. 

```
 symbol lookup error: ./TestLSTM-NNPA_main_graph.so: undefined symbol: zdnn_init
```

Reply:
Is the main branch of your PR up to date? After commit `9cd5cbe7f1fd1c729c9392de13d1a027be51a47f`, functions in `libzdnn.a` should all be included in the generated .so if the .so is created with the `--maccel=NNPA` option. You can add the `-v` option to see the commands for how `TestLSTM-NNPA_main_graph.so` is generated.

Reply:
I just tried your branch and I can see `zdnn_init` in the `TestLSTM-NNPA_main_graph.so`:
```
linux05:/tmp/onnx-mlir/build# nm ./test/accelerators/NNPA/numerical/TestLSTM-NNPA_main_graph.so|grep zdnn_init
0000000000013098 T zdnn_init
000000000000f930 T zdnn_init_pre_transformed_desc
0000000000017b70 T zdnn_init_ztensor
0000000000017c00 T zdnn_init_ztensor_with_malloc
```

Reply:
@gongsu832 Thanks for the test. Sorry, LSTM tests in current PR is no problem since all test configurations generate code for NNPA. It seems the error happens in the configurations which is not supported in zdnn LSTM. Following is such configuration. (`isNoneP` is 0)  In this configuration, we don't need to call zdnn_init, but I think it is ok to call with zdnn library. 

```
diff --git a/test/accelerators/NNPA/numerical/TestLSTM-NNPA.cpp b/test/accelerators/NNPA/numerical/TestLSTM-NNPA.c
pp
index b8bd4125..8dae1c4d 100644
--- a/test/accelerators/NNPA/numerical/TestLSTM-NNPA.cpp
+++ b/test/accelerators/NNPA/numerical/TestLSTM-NNPA.cpp
@@ -87,14 +87,14 @@ int main(int argc, char *argv[]) {
     const auto isNoneC = *rc::gen::element(0, 1);
     // Whether the weight tensor for peepholes(P) is specified.
     // Set none for peephole since zDNN does not support peephole
-    const auto isNoneP = 1;
+    const auto isNoneP = 0;
 
     RC_ASSERT(isOMLSTMTheSameAsNaiveImplFor(D, S, B, I, H, isDynS == 0,
         isDynB == 0, isNoneH == 1, isNoneC == 1, isNoneP == 1));
   });
   if (!success)
     return 1;
-
+  /*  Skip following tests for reducing test cases.
   // Exhaustive test case generation.
   // Set none for peephole
   for (int64_t s = 3; s < 4; s++)
@@ -115,6 +115,6 @@ int main(int argc, char *argv[]) {
                   // assert(isOMLSTMTheSameAsNaiveImplFor(
                   //    2, s, b, i, h, dyns, dynb, noneh, nonec, 1));
                 }
-
+  */
   return 0;
```

- I used `RC_PARAM=max_success=1` for reducing the tests.
```
diff --git a/test/accelerators/NNPA/numerical/CMakeLists.txt b/test/accelerators/NNPA/numerical/CMakeLists.txt
index a24d45d9..5c39eef3 100644
--- a/test/accelerators/NNPA/numerical/CMakeLists.txt
+++ b/test/accelerators/NNPA/numerical/CMakeLists.txt
@@ -36,11 +36,12 @@ function(add_numerical_unittest test_name)
 
   # Optimization level set by ONNX_MLIR_TEST_OPTLEVEL, defaults to 3
   add_test(NAME ${test_name}
-    COMMAND ${test_name} -O${ONNX_MLIR_TEST_OPTLEVEL} --maccel=NNPA --instrument-zhigh-ops=ALL --InstrumentBefore
ZHighOp --InstrumentAfterZHighOp --InstrumentReportTimeZHigh
+    COMMAND ${test_name} -O${ONNX_MLIR_TEST_OPTLEVEL} --maccel=NNPA --instrument-zhigh-ops=ALL --InstrumentBefore
ZHighOp --InstrumentAfterZHighOp --InstrumentReportTimeZHigh -v
     WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
     )
   set_tests_properties(${test_name} PROPERTIES LABELS numerical-nnpa)
-  set_tests_properties(${test_name} PROPERTIES ENVIRONMENT "TEST_ATOL=0.01;TEST_RTOL=0.05")
+  set_tests_properties(${test_name} PROPERTIES ENVIRONMENT "TEST_ATOL=0.01;TEST_RTOL=0.05;RC_PARAMS=max_success=1")

```

- I saw following output 

```
1/1 Test #1: TestLSTM-NNPA ....................***Failed    0.36 sec
Model can reuse the current seed by exporting "TEST_SEED=147355436"
Target options: "--maccel=NNPA"
Using configuration: max_success=1 seed=13632536526712692457

- LSTM implementation correctness
[/home/imaihal/work/onnx-mlir/build/test/accelerators/NNPA/numerical/]/home/imaihal/work/llvm-project/build/bin/op
t: opt -O3 --mtriple=s390x-ibm-linux -o ./TestLSTM-NNPA_main_graph.bc ./TestLSTM-NNPA_main_graph.unoptimized.bc
[/home/imaihal/work/onnx-mlir/build/test/accelerators/NNPA/numerical/]/home/imaihal/work/llvm-project/build/bin/ll
c: llc -O3 --mtriple=s390x-ibm-linux -filetype=obj -relocation-model=pic -o ./TestLSTM-NNPA_main_graph.o ./TestLST
M-NNPA_main_graph.bc
[/home/imaihal/work/onnx-mlir/build/test/accelerators/NNPA/numerical/]/usr/bin/c++: c++ ./TestLSTM-NNPA_main_graph.o -o ./TestLSTM-NNPA_main_graph.so -shared -fPIC -L/home/imaihal/docker/imaihal-ubuntu/work/onnx-mlir/build/Debug/lib -lcruntime -lzdnn -lRuntimeNNPA
/home/imaihal/work/onnx-mlir/build/Debug/bin/TestLSTM-NNPA: symbol lookup error: ./TestLSTM-NNPA_main_graph.so: undefined symbol: zdnn_init
```

```
$ nm ./build/test/accelerators/NNPA/numerical/TestLSTM-NNPA_main_graph
.so|grep zdnn
                 U zdnn_init
                 U zdnn_is_version_runnable
```

Reply:
@imaihal PR https://github.com/onnx/onnx-mlir/pull/1430 should fix this.

Reply:
@imaihal just landed, so update to the latest and you should have that fix

Reply:
@gongsu832 @AlexandreEichenberger Order! I confirmed it works well. Thanks!  I edited ` addCompilerConfig`, but I didn't change the order. Is there any specific reason why the order is affected?

Reply:
This is how dynamic linking works. libRuntimeNNPA has functions that call `zdnn_init` so libRuntimeNNPA depends on libzdnn and the linkage order must be `-lRuntimeNNPA -lzdnn`.

If you have `-lzdnn -lRuntimeNNPA`, when the linker sees the calls to `zdnn_init` in libRuntimeNNPA, it doesn't have any libraries after it to resolve `zdnn_init` so it will mark it as "U" (Unresolved) as you saw above. If this were part of generating an executable, the linker would have failed and produced an unresolved symbol error immediately. However, in this case we are generating a .so (`TestLSTM-NNPA_main_graph.so`) so the linker tolerates unresolved symbols because you can still resolve them at runtime through LD_PRELOAD or LD_LIBRARY_PATH.

Reply:
I see. It seemed some configurations worked in the original order (`-lRuntimeNNPA -lzdnn`), but did they happen to work even in wrong order?

Reply:
The original wrong order before the PR fix was `-lzdnn -lRuntimeNNPA`. Some tests still worked probably because some function in the model runtime might have called `zdnn_init`, i.e., the command to create the .so is something like
```
c++ model.o -o model.so -shared -fPIC -L/opt/Downloads/onnx-mlir/build/Release/lib -lzdnn -lRuntimeNNPA -lcruntime
```
So if some function in `model.o` calls `zdnn_init` then it would have been resolved by `-lzdnn` even if `-lRuntimeNNPA` is in the wrong order.

Reply:
I could clearly see. Thank you very much!